client/lib/registration.js

```diff
@@ -0,0 +1,257 @@
+        /*
+         *   - locate uuid on disk
+         *   - sign uuid with private key
+         *   - send login request
+         */
+        logger.info('We have keys already');
+      }
+      else {
+        if (!self.generateKeys()) {
+          return reject('Failed to generate keys');
+        }
+        if (!self.saveKeysSync()) {
+          return reject('Failed to save keys to disk');
+        }
+        logger.info('Creating registration..');
+        api.create({
+          pubKey: self.getPublicKey(),
+        }).then((res) => {
+          logger.info('Registration create:', res);
+          self.uuid = res.uuid;
+          if (!self.saveUUIDSync()) {
+            reject('Failed to save UUID!');
+          }
+          else {
+            resolve(res);
+          }
+        }).catch((res) => {
+          logger.error('Failed to register:', res);
+          reject(res);
+        });
+      }
+    });
+  }
+
+  /*
+   * Generate the keys necessary for this instance
+   *
+   * @return Boolean
+   */
+  generateKeys() {
+    let ecdh = crypto.createECDH('secp521r1');
+    let pubKey = ecdh.generateKeys('base64');
+    logger.info('Generated public key:', pubKey);
+
+    this.publicKey = pubKey;
+    this.privateKey = ecdh.getPrivateKey('base64');
+
+    /* If we have a private key, that's good enough! */
+    if (this.privateKey) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /*
+   * Return the base64 encoded public key associated with this instance
+   *
+   * Will return null if no public key has yet been generated
+   *
+   * @return String
+   */
+  getPublicKey() {
+    return this.publicKey;
+  }
+
+  /*
+   * Persist the UUID generated by the call to the registration service
+   *
+   * @return Boolean on success
+   */
+  saveUUIDSync() {
+    if (this.uuid != null) {
+      const config = {
+        uuid: this.uuid,
+      };
+      logger.info('Saving uuid to disk at', this.uuidPath());
+
+      fs.writeFileSync(
+        this.uuidPath(),
+        JSON.stringify(config),
+        this.fileOptions);
+      return true;
+    }
+    return false;
+  }
+
+  /*
+   * Persist generated keys
+   *
+   * @return Boolean on success
+   */
+  saveKeysSync() {
+    if ((!this.publicKey) || (!this.privateKey)) {
+      logger.warn('saveKeysSync() called without a private or public key on Registration');
+      return false;
+    }
+
+    const keyPath = this.keyPath();
+    const publicKeyPath = this.publicKeyPath();
+    const privateKeyPath = this.privateKeyPath();
+
+    logger.debug('Writing public key to', publicKeyPath);
+    fs.writeFileSync(publicKeyPath, this.publicKey, this.fileOptions);
+
+    logger.debug('Writing private key to', privateKeyPath);
+    fs.writeFileSync(privateKeyPath, this.privateKey, this.fileOptions);
+
+    return true;
+  }
+
+  /*
+   * Load generated keys from disk assuming they are present
+   *
+   * @return Boolean on successful load of keys
+   */
+  loadKeysSync() {
```

Comment:
Is this called from anywhere but tests?

Reply:
This is not called yet because it's required for the Login flow, this code isn't quite there yet.